File: python/kmp.py


def prefix_table(s:str) -> list:
	"""return the prefix table of string s"""
	# print(len(s))
	assert len(s) >= 2
	l = [-1, 0]
	slow, fast = 0, 1
	while fast < len(s) - 1:
		if s[slow] == s[fast]:
			slow += 1
			l.append(slow)
			fast += 1
		elif slow > 0:
			slow = l[slow]
		else:
			l.append(0)
			fast += 1
	return l

print(prefix_table("ababcaccabbbbbbbbbab"))
# 输出[-1, 0, 0, 1, 2, 0, 1, 0, 0, 1, 2, 0, 0, 0, 0, 0, 0, 0, 0, 1]

File: java/kmp.java

public class KMP {
	
	/* Return the prefix table of string s */
	public static int[] prefix_table(String s) {
		int[] next = new int[s.length()];
		next[0] = -1;
		int j = 0, k = -1;
		while (j < s.length() - 1) {
			if (k == -1 || s.charAt(j) == s.charAt(k)) {
				next[++j] = ++k;
			} else {
				k = next[k];
			}
		}
		return next;
	}
	
	/* Return the first index if the pattern matches the target else -1 */
	public static int kmp(String target, String pattern) {
		int[] next = prefix_table(pattern);
		int t = 0, p = 0; // points to the target and the pattern repectively.
		while (t < target.length() && p < pattern.length()) {
			if (p == -1 || target.charAt(t) == pattern.charAt(p)) {
				p++;
				t++;
			} else {
				p = next[p];
			}
		}
		if (p == pattern.length()) {
			return t - p;
		} else {
			return -1;
		}
	}
	
}
			
		
}